feat(website): Azure cost-post (#5940)
cloudquery/cloudquery: website/pages/blog/azure-cost-optimization-with-cloudquery.md

---
title: Correlating Costs to Resources for Azure Cost Optimization with CloudQuery
tag: tutorial

> date: 2022/12/23

Author: This date should be edited to reflect the date of the blog, and should come after the Azure plugin release that contains https://github.com/cloudquery/cloudquery/pull/5898

Author: ... which will be azure v3.1.0

First, we need to create a _Cost Analysis Report_ in Azure. To do that, go to the Cost Management section in the Azure Portal and click on the [Cost Analysis](https://portal.azure.com/#view/Microsoft_Azure_CostManagement/Menu/~/costanalysis/openedBy/AzurePortal) tab.

![Azure Cost Analysis Start Page](/images/blog/azure-cost-optimization-with-cloudquery/cost-analysis-cost-by-resource-report.png)

Then, click on the _Cost by Resource_ button.

![Azure Cost By Resource Menu](/images/blog/azure-cost-optimization-with-cloudquery/cost-analysis-cost-by-resource-menu.png)

Here you can make adjustments to your report. Make sure to select the time range you want to analyze. For this tutorial, we'll leave this at the default of current billing period. Make sure the _Group By_ option is set to **Resource**.

![Azure Cost By Resource Report](/images/blog/azure-cost-optimization-with-cloudquery/cost-analysis-cost-by-resource-report-annotated.png)

Use the _Save as_ button to save this as a new report view. Give it a name and click _Save_.

![Azure Cost By Resource Report Save Button](/images/blog/azure-cost-optimization-with-cloudquery/cost-analysis-cost-by-resource-report-save.png)

> From now on every time we run a `cloudquery sync` command, the data from this report will be synced to PostgreSQL database. 

Reviewer: ```suggestion
Now we can configure CloudQuery so each time we run `cloudquery sync` the data from this report will be synced to a PostgreSQL database.
```

Not sure, but technically isn't `From now on every time` is only correct after following the next steps?

Author: Yes better, thanks :)

## Syncing data

> Before going off and writing queries, we need to sync the data from Azure to our PostgreSQL database. To do that, we'll use the `cloudquery sync` command. Let's start with writing (well, copy-pasting) a configuration file (For full config reference, checkout the [Azure Source Plugin](https://www.cloudquery.io/docs/plugins/sources/azure/configuration) and [PostgreSQL Destination Plugin](https://www.cloudquery.io/docs/plugins/destinations/postgresql/overview)):

Reviewer: ```suggestion
Before going off and writing queries, we need to sync the data from Azure to our PostgreSQL database. To do that, we'll use the `cloudquery sync` command. Let's start with creating a configuration file (For the full configuration reference, check out the [Azure Source Plugin](https://www.cloudquery.io/docs/plugins/sources/azure/configuration) and [PostgreSQL Destination Plugin](https://www.cloudquery.io/docs/plugins/destinations/postgresql/overview)):
```

```yaml copy
kind: source
spec:
  # Source spec section
  name: "azure"
  path: "cloudquery/azure"
  version: "VERSION_SOURCE_AZURE"
  destinations: ["postgresql"]

>   concurrency: 100 # Azure is a beast: If you're running this on a beefy machine, you can increase this

Reviewer: ```suggestion
  concurrency: 100 # The Azure source plugin supports many resources and can fail on small machines. Increase this if you run it on a beefy machine
```

>   tables: ["*"] # "azure_costmanagement_views" and "azure_costmanagement_view_queries" are required for this tutorial, but let's get them all
>   spec:
>     # We could specify a list of subscriptions to sync, but we'll just sync all of them
>     # subscriptions: ["00000000-0000-0000-0000-000000000000"]

Reviewer: ```suggestion
  tables: ["*"] # "azure_costmanagement_views" and "azure_costmanagement_view_queries" are required for this tutorial, but let's get them all
```

For brevity. Also maybe we can sync just the required table so we can drop the `concurrency` configuration

Author: If we just sync the required tables then there's no way of correlating them with other data later, because the rows won't exist.

### Deeper Dive

After seeing actual data it will become apparent that we'll need to figure out the column indexes for the data we're interested in. We can do this by looking at the `properties.columns` data in the `azure_costmanagement_view_queries` table. We can use `JSONB_ARRAY_ELEMENTS WITH ORDINALITY` to match column names to indexes. Something like `JSONB_ARRAY_ELEMENTS(q.properties->'columns') WITH ORDINALITY cid(col, pos) ON cid.col->>'name'='ResourceId'` will give us the index of the ResourceId column as `cid.pos`. A full query to return _Resource Id_ vs. _Cost (USD)_ might look like:

```sql copy 
SELECT (r->>(cid.pos::int - 1))::text AS res_id, (r->>(ccost.pos::int - 1))::numeric AS cost_usd
FROM azure_costmanagement_view_queries q
JOIN JSONB_ARRAY_ELEMENTS(q.properties->'rows') AS r ON TRUE
JOIN JSONB_ARRAY_ELEMENTS(q.properties->'columns') WITH ORDINALITY cid(col, pos) ON cid.col->>'name'='ResourceId'
JOIN JSONB_ARRAY_ELEMENTS(q.properties->'columns') WITH ORDINALITY ccost(col, pos) ON ccost.col->>'name'='CostUSD'
JOIN azure_costmanagement_views v ON v._cq_id=q._cq_parent_id
WHERE v.name='cost-matching'
```

> Looks complicated, but it's not too bad. Let's break it down:

Reviewer: ```suggestion
Let's break it down:
```

By writing `complicated` we're already planting the idea it's complicated.
